pyccl/pk2d.py

```diff
@@ -254,56 +321,52 @@ def eval(self, k, a, cosmo=None, *, derivative=False):
             self.psp.extrap_linear_growth = 401  # flag extrapolation
 
-        status = 0
-        if isinstance(k, int):
-            k = float(k)
-        if isinstance(k, float):
-            f, status = eval_funcs[0](self.psp, np.log(k), a,
-                                      cosmo.cosmo, status)
-        else:
-            k_use = np.atleast_1d(k)
-            f, status = eval_funcs[1](self.psp, np.log(k_use), a,
-                                      cosmo.cosmo, k_use.size, status)
-
-        # Catch scale factor extrapolation bounds error.
-        if status == lib.CCL_ERROR_SPLINE_EV:
-            raise TypeError(
-                "Pk2D evaluation scale factor is outside of the "
-                "interpolation range. To extrapolate, pass a Cosmology.")
-        check(status, cosmo)
-        return f
+        a_use = np.atleast_1d(a).astype(float)
+        k_use = np.atleast_1d(k).astype(float)
+        lk_use = np.log(k_use)
 
-    # Save a dummy cosmology as an attribute of the `eval` method so we don't
-    # have to initialize one every time no `cosmo` is passed. This is gentle
-    # with memory too, as `free` does not work for an empty cosmology.
-    eval._cosmo = type("Dummy", (object,), {"cosmo": lib.cosmology()})()
-
-    def eval_dlogpk_dlogk(self, k, a, cosmo):
-        """Evaluate logarithmic derivative. See ``Pk2D.eval`` for details."""
-        return self.eval(k, a, cosmo=cosmo, derivative=True)
-
-    def __call__(self, k, a, cosmo=None, *, derivative=False):
-        """Callable vectorized instance."""
-        out = np.array([self.eval(k, aa, cosmo=cosmo, derivative=derivative)
-                        for aa in np.atleast_1d(a).astype(float)])
-        return out.squeeze()[()]
+        status = 0
+        out = np.zeros([len(a_use), len(k_use)])
+        for ia, aa in enumerate(a_use):
+            f, status = eval_func(self.psp, lk_use, aa,
+                                  cosmo.cosmo, k_use.size, status)
+
+            # Catch scale factor extrapolation bounds error.
+            if status == lib.CCL_ERROR_SPLINE_EV:
+                raise TypeError(
+                    "Pk2D evaluation scale factor is outside of the "
+                    "interpolation range. To extrapolate, pass a Cosmology.")
+            check(status, cosmo)
+            out[ia, :] = f
+
+        if np.ndim(k) == 0:
+            out = np.squeeze(out, axis=-1)
+        if np.ndim(a) == 0:
+            out = np.squeeze(out, axis=0)
+        return out
+
+    # Save a dummy cosmology as an attribute of the `__call__` method
+    # so we don't have to initialize one every time no `cosmo` is passed.
+    # This is gentle with memory too, as `free` does not work for an empty
+    # cosmology.
+    __call__._cosmo = type("Dummy", (object,), {"cosmo": lib.cosmology()})()
 
     def copy(self):
         """Return a copy of this Pk2D object."""
         if not self:
-            return Pk2D(empty=True)
+            return CCLObject.__new__(Pk2D)
```

Comment:
Same here about `Pk2D.__new__`.

Reply:
as above, resolve if you don't think it's a problem.

Reply:
Done